Slug with explicit .html extension breaks build

### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [X] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

I'm migrating a blog from Jekyll. It is necessary for me to keep the exact same url structure because there's external refs to these posts. So I have a

    slug: a/b.html

in every blog post frontmatter. On build, this fails with the following error:

    [ERROR] Unable to build website for locale en.
    [ERROR] Error: Expected output HTML file to be found at /home/projects/kkviyzpmp.github/build/blog/a/b.html/index.html.

Running the site works fine, error only happens on build.

### Reproducible demo

https://stackblitz.com/edit/github-iuz7vq?file=blog/2021-08-26-welcome/index.md

### Steps to reproduce

1. Set a slug in a blog post frontmatter: `/a/b.html`
2. Run `yarn build`

### Expected behavior

Build should succeed, especially since running the site poses no problems.

### Actual behavior

Build fails with the following error message:

    [ERROR] Unable to build website for locale en.
    [ERROR] Error: Expected output HTML file to be found at /home/projects/kkviyzpmp.github/build/blog/a/b.html/index.html.


### Your environment

On stackblitz playground.

### Self-service

- [ ] I'd be willing to fix this bug myself.

@rwwilden please when. you have a stacktrace, post it in its entirety. You removed the stacktrace from the error message, preventing us to immediately understand what's the issue, needing to reproduce it locally to understand the problem.

The stacktrace is:

```
[ERROR] Unable to build website for locale en.
[ERROR] Error: Expected output HTML file to be found at /Users/sebastienlorber/Desktop/projects/docusaurus/website/build/blog/docu2.1.html/index.html.
    at readOutputHTMLFile (/Users/sebastienlorber/Desktop/projects/docusaurus/packages/docusaurus-utils/lib/emitUtils.js:75:15)
    at async /Users/sebastienlorber/Desktop/projects/docusaurus/packages/docusaurus-plugin-content-blog/lib/feed.js:41:25
    at async Promise.all (index 0)
    at async generateBlogFeed (/Users/sebastienlorber/Desktop/projects/docusaurus/packages/docusaurus-plugin-content-blog/lib/feed.js:39:5)
    at async createBlogFeedFiles (/Users/sebastienlorber/Desktop/projects/docusaurus/packages/docusaurus-plugin-content-blog/lib/feed.js:85:18)
    at async Object.postBuild (/Users/sebastienlorber/Desktop/projects/docusaurus/packages/docusaurus-plugin-content-blog/lib/index.js:315:13)
    at async /Users/sebastienlorber/Desktop/projects/docusaurus/packages/docusaurus/lib/commands/build.js:171:9
    at async Promise.all (index 1)
    at async buildLocale (/Users/sebastienlorber/Desktop/projects/docusaurus/packages/docusaurus/lib/commands/build.js:167:5)
    at async tryToBuildLocale (/Users/sebastienlorber/Desktop/projects/docusaurus/packages/docusaurus/lib/commands/build.js:38:20)
```

This makes it clear for us maintainers that the problem is in our blog feed: we added a `readOutputHTMLFile()` method to create the blog RSS feed HTML content, but it's not reading from the correct output file location.

When an URL already ends with `.html`, such as `/x/y/z.html`, we output `/x/y/z.html` and not `/x/y/z/index.html` => `readOutputHTMLFile` apparently doesn't support this logic

--- a/packages/docusaurus-utils/src/__tests__/__fixtures__/build-snap/htmlFile.html
+++ b/packages/docusaurus-utils/src/__tests__/__fixtures__/build-snap/htmlFile.html
@@ -1,0 +1,1 @@
+htmlFile.html

--- a/packages/docusaurus-utils/src/__tests__/__fixtures__/build-snap/weird.html.folder/nestedHtmlFile.html
+++ b/packages/docusaurus-utils/src/__tests__/__fixtures__/build-snap/weird.html.folder/nestedHtmlFile.html
@@ -1,0 +1,1 @@
+nestedHtmlFile.html

--- a/packages/docusaurus-utils/src/__tests__/emitUtils.test.ts
+++ b/packages/docusaurus-utils/src/__tests__/emitUtils.test.ts
@@ -89,6 +89,22 @@ describe('readOutputHTMLFile', () => {
       ).then(String),
     ).resolves.toBe('htmlFile.html\n');
   });
+  it('reads file ending in .html in folder containing .html', async () => {
+    await expect(
+      readOutputHTMLFile(
+        '/weird.html.folder/nestedHtmlFile.html',
+        path.join(__dirname, '__fixtures__/build-snap'),
+        undefined,
+      ).then(String),
+    ).resolves.toBe('nestedHtmlFile.html\n');
+    await expect(
+      readOutputHTMLFile(
+        '/weird.html.folder/nestedHtmlFile.html',
+        path.join(__dirname, '__fixtures__/build-snap'),
+        undefined,
+      ).then(String),
+    ).resolves.toBe('nestedHtmlFile.html\n');
+  });
   // Can it ever happen?
   it('throws if file does not exist', async () => {
     await expect(
@@ -98,7 +114,7 @@ describe('readOutputHTMLFile', () => {
         undefined,
       ).then(String),
     ).rejects.toThrowErrorMatchingInlineSnapshot(
-      `"Expected output HTML file to be found at <PROJECT_ROOT>/packages/docusaurus-utils/src/__tests__/__fixtures__/build-snap/nonExistent/index.html."`,
+      `"Expected output HTML file to be found at <PROJECT_ROOT>/packages/docusaurus-utils/src/__tests__/__fixtures__/build-snap/nonExistent/index.html for permalink /nonExistent."`,
     );
   });
 });

--- a/packages/docusaurus-utils/src/__tests__/emitUtils.test.ts
+++ b/packages/docusaurus-utils/src/__tests__/emitUtils.test.ts
@@ -73,6 +73,22 @@ describe('readOutputHTMLFile', () => {
       ).then(String),
     ).resolves.toBe('file\n');
   });
+  it('reads file ending in .html', async () => {
+    await expect(
+      readOutputHTMLFile(
+        '/htmlFile.html',
+        path.join(__dirname, '__fixtures__/build-snap'),
+        false,
+      ).then(String),
+    ).resolves.toBe('htmlFile.html\n');
+    await expect(
+      readOutputHTMLFile(
+        '/htmlFile.html',
+        path.join(__dirname, '__fixtures__/build-snap'),
+        undefined,
+      ).then(String),
+    ).resolves.toBe('htmlFile.html\n');
+  });
   // Can it ever happen?
   it('throws if file does not exist', async () => {
     await expect(

--- a/packages/docusaurus-utils/src/emitUtils.ts
+++ b/packages/docusaurus-utils/src/emitUtils.ts
@@ -79,20 +79,22 @@ export async function readOutputHTMLFile(
   trailingSlash: boolean | undefined,
 ): Promise<Buffer> {
   const withTrailingSlashPath = path.join(outDir, permalink, 'index.html');
-  const withoutTrailingSlashPath = path.join(
-    outDir,
-    `${permalink.replace(/\/$/, '').replace(/.html$/, '')}.html`,
-  );
-  const HTMLPath = await findAsyncSequential(
-    [
-      trailingSlash !== false && withTrailingSlashPath,
-      trailingSlash !== true && withoutTrailingSlashPath,
-    ].filter((p): p is string => Boolean(p)),
-    fs.pathExists,
-  );
+  const withoutTrailingSlashPath = (() => {
+    const basePath = path.join(outDir, permalink.replace(/\/$/, ''));
+    const htmlSuffix = /\.html?$/i.test(basePath) ? '' : '.html';
+    return `${basePath}${htmlSuffix}`;
+  })();
+
+  const possibleHtmlPaths = [
+    trailingSlash !== false && withTrailingSlashPath,
+    trailingSlash !== true && withoutTrailingSlashPath,
+  ].filter((p): p is string => Boolean(p));
+
+  const HTMLPath = await findAsyncSequential(possibleHtmlPaths, fs.pathExists);
+
   if (!HTMLPath) {
     throw new Error(
-      `Expected output HTML file to be found at ${withTrailingSlashPath}.`,
+      `Expected output HTML file to be found at ${withTrailingSlashPath} for permalink ${permalink}.`,
     );
   }
   return fs.readFile(HTMLPath);

--- a/packages/docusaurus-utils/src/emitUtils.ts
+++ b/packages/docusaurus-utils/src/emitUtils.ts
@@ -81,7 +81,7 @@ export async function readOutputHTMLFile(
   const withTrailingSlashPath = path.join(outDir, permalink, 'index.html');
   const withoutTrailingSlashPath = path.join(
     outDir,
-    `${permalink.replace(/\/$/, '')}.html`,
+    `${permalink.replace(/\/$/, '').replace(/.html$/, '')}.html`,
   );
   const HTMLPath = await findAsyncSequential(
     [

--- a/website/_dogfooding/_blog tests/2022-10-02-html-slug.md
+++ b/website/_dogfooding/_blog tests/2022-10-02-html-slug.md
@@ -1,0 +1,15 @@
+---
+title: A post with html slug
+tags: [paginated-tag]
+slug: /x/y/z.html
+---
+
+# Hmmm!
+
+This is a blog post with an html slug!
+
+```mdx-code-block
+import Partial from "./_partial.mdx"
+
+<Partial />
+```

